File: marcel-stdlib/src/main/java/marcel/lang/URLMarcelScriptLoader.java
package marcel.lang;

import java.io.File;
import java.net.URLClassLoader;

public class URLMarcelScriptLoader extends MarcelScriptLoader {

  private final ClassLoader parentLoader;

  public URLMarcelScriptLoader(ClassLoader parentLoader) {
    this.parentLoader = parentLoader;
  }

  public URLMarcelScriptLoader() {
    this(URLMarcelScriptLoader.class.getClassLoader());
  }

  @Override
  public Script loadScript(String className, File jarFile, Binding binding) throws ReflectiveOperationException {
    if (!jarFile.isFile()) throw new IllegalArgumentException(String.format("File %s is not a regular file", jarFile));
    // load the jar into the classpath
    URLClassLoader classLoader = new URLClassLoader(getJarUrls(jarFile), parentLoader);

    // and then load the class
    Class<?> clazz = classLoader.loadClass(className);
    if (!Script.class.isAssignableFrom(clazz)) {
      throw new IllegalArgumentException("The loaded class is not a script");
    }
    if (binding == null) {
      return (Script) clazz.getDeclaredConstructor().newInstance();
    } else {
      return (Script) clazz.getDeclaredConstructor(Binding.class).newInstance(binding);
    }
  }
}

File: subprojects/marcel-repl/src/main/kotlin/com/tambapps/marcel/repl/URLMarcelScriptLoader.kt
package com.tambapps.marcel.repl

import marcel.lang.Binding
import marcel.lang.Script
import java.io.File
import java.net.URLClassLoader
import kotlin.jvm.Throws

class URLMarcelScriptLoader(
  private val parentLoader: ClassLoader
): MarcelScriptLoader() {

  constructor(): this(URLMarcelScriptLoader::class.java.classLoader)

  @Throws(ClassNotFoundException::class)
  override fun loadScript(className: String, jarFile: File, binding: Binding?): Script {
    if (!jarFile.isFile) throw IllegalArgumentException("File $jarFile is not a regular file")

    // load the jar into the classpath
    val classLoader = URLClassLoader(getJarUrls(jarFile), parentLoader)
    // and then load the class
    val clazz = classLoader.loadClass(className)
    if (!Script::class.java.isAssignableFrom(clazz)) {
      throw IllegalArgumentException("The loaded class is not a script")
    }
    return if (binding == null) clazz.getDeclaredConstructor().newInstance() as Script
    else clazz.getDeclaredConstructor(Binding::class.java).newInstance(binding) as Script
  }

}